asynction/security.py

```diff
@@ -0,0 +1,412 @@
+    try:
+        lhs, rhs = authorization.split(None, 1)
+        return lhs, rhs
+    except ValueError as err:
+        raise SecurityException from err
+
+
+def validate_basic(
+    request: Request, basic_info_func: BasicInfoFunc, required_scopes: Sequence[str]
+) -> Union[Mapping, None]:
+    auth = extract_auth_header(request)
+    if not auth:
+        return None
+    auth_type, user_pass = auth
+    if not auth_type or not user_pass:
+        raise SecurityException
+
+    if HTTPAuthenticationScheme(auth_type.lower()) != HTTPAuthenticationScheme.BASIC:
+        return None
+
+    try:
+        username, password = base64.b64decode(user_pass).decode("latin1").split(":", 1)
+    except Exception as err:
+        raise SecurityException from err
+
+    if not username or not password:
+        raise SecurityException
+
+    token_info = basic_info_func(username, password, required_scopes)
+    if token_info is None:
+        raise SecurityException
+
+    return token_info
+
+
+def validate_authorization_header(
+    request: Request, token_info_func: TokenInfoFunc
+) -> Union[Mapping, None]:
+    """Check that the provided request contains a properly formatted Authorization
+    header and invokes the token_info_func on the token inside of the header.
+    """
+    auth = extract_auth_header(request)
+    if not auth:
+        return None
+    auth_type, token = auth
+    if not auth_type or not token:
+        raise SecurityException
+
+    if auth_type.lower() != "bearer":
+        return None
+
+    token_info = token_info_func(token)
+    if token_info is None:
+        raise SecurityException
+
+    return token_info
+
+
+def validate_api_key(
+    request: Request,
+    api_key_info_func: APIKeyInfoFunc,
+    required_scopes: Sequence[str],
+    bearer_format: Optional[str] = None,
+) -> Union[Mapping, None]:
+    """
+    Adapted from: https://github.com/zalando/connexion/blob/main/connexion/security/security_handler_factory.py#L221  # noqa: 501
+    """
+    auth = extract_auth_header(request)
+    if not auth:
+        return None
+    auth_type, token = auth
+
+    if not auth_type or not token:
+        raise SecurityException
+
+    if auth_type.lower() != "bearer":
+        return None
+
+    token_info = api_key_info_func(token, required_scopes, bearer_format)
+    if token_info is None:
+        raise SecurityException
+
+    return token_info
+
+
+def validate_scopes(
+    required_scopes: Sequence[str], token_scopes: Sequence[str]
+) -> bool:
+    """Validates that all require scopes are present in the token scopes"""
+    missing_scopes = set(required_scopes) - set(token_scopes)
+    if missing_scopes:
+        raise SecurityException(f"Missing required scopes: {missing_scopes}")
+
+    return not missing_scopes
+
+
+def load_scope_validate_func(scheme: SecurityScheme) -> ScopeValidateFunc:
+    # importing here because doing it at the top leads to a circular import
+    from asynction.server import load_handler
```

Comment:
Oh, I only now realised the circular dependency issue.

We should move this to a new `uitls.py` module I guess, which is what you did in the original PR (sorry).